File: NeetCodeIO/78Subsets/Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # Backtracking -> Time:O(2^n), Space:O(n)
        ans = []
        def dfs(i, sub):
            if i >= len(nums):
                ans.append(sub.copy())
                return
            dfs(i+1, sub.copy())
            sub.append(nums[i])
            dfs(i+1, sub.copy())
        dfs(0, [])
        return ans

File: NeetCodeIO/78Subsets/Subsets.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void dfs(vector<int>& nums, vector<int> sub, int ind) {
        if (ind >= nums.size()) {
            ans.push_back(sub);
            return;
        }
        dfs(nums, sub, ind+1);
        sub.push_back(nums[ind]);
        dfs(nums, sub, ind+1);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        // Backtracking -> Time:O(2^n), Space:O(n)
        vector<int> subSet;
        dfs(nums, subSet, 0);
        return ans;
    }
};
